src/main/java/com/facebook/openwifirrm/modules/Modeler.java

```diff
@@ -271,7 +272,7 @@ private void fetchInitialData() {
 					State stateModel = gson.fromJson(state, State.class);
 					dataModel.latestStates.computeIfAbsent(
 						device.serialNumber,
-						k -> new ArrayList<>()
+						k -> Collections.synchronizedList(new ArrayList<>())
```

Comment:
AFAICT, probably the best option. COW is going to be too expensive.

Reply:
also maybe not in this PR but i think there are no accesses besides first and last usually (besides sorting the whole thing) - linked list might make more sense vs allocating contiguous block